### PR TITLE
fix(middleware/persist): ensure `persist` does not drop updates in `onRehydrateStorage` when using a synchronous storage API

### LINKED ISSUE #1688: State updates within the 'onRehydrateStorage' callback using a synchronous storage API get dropped
## Summary

Updating a store's state within the `onRehydrateStorage` callback when using a synchronous storage API results in the state update getting overwritten by the persisted state. The overwritten state update will still get persisted, however, resulting in a scenario where (1) the update failed to be applied to the current in-memory store and (2) the in-memory state and the persisted state do not match.

This issue is caused by [the return value of `createState`](https://github.com/pmndrs/zustand/blob/962f83e01599257aa6eceaaa993965a0e296d359/src/vanilla.ts#L101) (which in the synchronous case is [`stateFromStorage`](https://github.com/pmndrs/zustand/blob/962f83e01599257aa6eceaaa993965a0e296d359/src/middleware/persist.ts#L347)) overwriting [state updates within the 'onRehydrateStorage' hook](https://github.com/pmndrs/zustand/blob/962f83e01599257aa6eceaaa993965a0e296d359/src/middleware/persist.ts#L303). The in-callback updates do _initially_ update the current state (and get persisted) via [`set` and `setItem`](https://github.com/pmndrs/zustand/blob/962f83e01599257aa6eceaaa993965a0e296d359/src/middleware/persist.ts#L398-L399), but they then get lost because [`stateFromStorage` is never updated to reflect those changes](https://github.com/pmndrs/zustand/blob/962f83e01599257aa6eceaaa993965a0e296d359/src/middleware/persist.ts#L294-L306).

This doesn't occur in the asynchronous case because the `createState` function immediately returns, setting the initial state to `undefined`. Then, when the hydration promises resolve during subsequent event loop ticks, they never get overwritten because the `createState` method already returned.

I believe this is also the root cause of #1527 - I tested that issue (using the sandbox provided) without `partialize` defined but it still had the same problem.

## Link to reproduction

I forked the sandbox from #1527 and isolated the issue described above. Feel free to toggle between (1) synchronously and asynchronously updating the state within `onRehydrateStorage` and (2) `localStorage` and `asyncStorage` to see how synchronous/asynchronous APIs affect the issue.

https://codesandbox.io/s/zustand-4-3-x-bug-forked-45m5jq?file=/src/App.js

## Check List

Please do not ask questions in issues.

- [X] I've already opened a [discussion](https://github.com/pmndrs/zustand/discussions) before opening this issue, or already discussed in other media.

Please include a minimal reproduction.

> N/A - I've got a reproducible example and opened a PR to fix.

- [X] I've added a link to a [typescript playground](https://www.typescriptlang.org/play) or [codesandbox](https://codesandbox.io) with a minimal reproduction.

Please check this if you're filing an issue regarding TypeScript.

- [X] I've read the [typescript guide](https://github.com/pmndrs/zustand/blob/main/docs/guides/typescript.md), in particular that `create` is to be used as `create<T>()(...)` and not `create<T>(...)`.

### PR DESCRIPTION
**Pull request by coffeebeats**
## Related Issues or Discussions

Fixes #1688

## Summary

See #1688 for a detailed description of the problem. This PR resolves the issue by ensuring `stateFromStorage` gets updated to the most recent value. I also added tests verifying this update pattern works for both synchronous and asynchronous storage APIs.

One note for reviewers: I added a comment describing how, in the asynchronous case, it's possible that the state passed to `postRehydrationCallback` is _not_ the most up-to-date state. This would occur if there was a state update between callbacks in the hydration promise chain. Please let me know if I should leave the comment as-is, make the change, or simply remove the comment.

## Check List

- [X] `yarn run prettier` for formatting code and docs

### CODE REVIEW

**Review by dai-shi — COMMENTED** [review]
looks correct.
can anyone else review too?

**Review by sewera — APPROVED** [review]
Looks correct also to me.

**Comment by sewera** [comment]
It's a tricky topic that will not have to be rediscovered with this comment, so I'd leave it as-is.

**Comment by dai-shi** [comment]
> Does this also need to be fixed for the [old implementation](https://github.com/pmndrs/zustand/blob/962f83e01599257aa6eceaaa993965a0e296d359/src/middleware/persist.ts#L303-L304)?

My plan is to leave the old implementation unchanged. People shouldn't use the deprecated one moving forward.

**Review by dai-shi — APPROVED** [review]

File: tests/persistSync.test.tsx [hunk 1]
```diff
@@ -549,6 +549,7 @@ describe('persist middleware with sync configuration', () => {
     expect(onFinishHydrationSpy2).toBeCalledWith({ count: 2 })
   })
 
+
```
Comment (dai-shi) [inline tests/persistSync.test.tsx:552]:
```suggestion
```

File: tests/persistSync.test.tsx [hunk 2]
```diff
@@ -598,5 +598,28 @@ describe('persist middleware with sync configuration', () => {
       { count: 42, name: 'test-storage' },
       undefined
     )
+
```
Comment (dai-shi) [inline tests/persistSync.test.tsx:601]:
```suggestion
  })

```

File: tests/persistAsync.test.tsx [hunk 3]
```diff
@@ -611,5 +611,40 @@ describe('persist middleware with async configuration', () => {
       { count: 42, name: 'test-storage' },
       undefined
     )
+
```
Comment (dai-shi) [inline tests/persistAsync.test.tsx:614]:
```suggestion
  })

```

**Pull request merged**